### PR TITLE
Camera support opaque texture  

### PR DESCRIPTION
**Pull request by GuoLei1990**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Major changes:
- Camera support opaque texture with `opaqueTextureEnabled ` and `opaqueTextureDownsampling `
- Camera support enable independent Canvas  by `independentCanvasEnabled`, enable this can on/off MSAA per camera by `msaaSamples `(opaqueTextureEnabled force turn on this mode)
- Add `BasicResources` class to store internal resource
- Remove internal function `_setRenderTargetInfo` to simplify the logic

### TODO:
- [x] Support camera viewport
- [x] Opt API
- [x] camera viewport
- [x] MSAA
- [x] Remove render pass
- [x] RenderTarget Test
- [x] recreateRenderTargetIfNeeded depthTexture waste
- [x] framebuffer 重复绑定
- [x] PerCamera MSAA with no OpaqueTexture

### Future:
- support blit scaleOffset
- Refactor Basic Resource
- ClearFlag(No Color) with invalidateFramebuffer
- SetActiveTarget viewport 可能存在重复设置浪费
-  Blit 优化(Triangle 模式等)

### CODE REVIEW

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 1]
```diff
@@ -166,59 +100,100 @@ export class BasicRenderPipeline {
...
       }
     } else {
-      camera.shaderData.setTexture(Camera._cameraDepthTextureProperty, camera.engine._whiteTexture2D);
+      camera.shaderData.setTexture(Camera._cameraDepthTextureProperty, engine._whiteTexture2D);
+    }
+
+    // Check if need to create internal color texture
+    const independentCanvasEnabled = camera.independentCanvasEnabled;
+    if (independentCanvasEnabled && Vector4.equals(camera.viewport, PipelineUtils.defaultViewport)) {
+      console.warn(
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:109]: use "Logger." in loop

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 2]
```diff
@@ -166,59 +100,100 @@ export class BasicRenderPipeline {
...
-            this._drawBackgroundTexture(engine, background);
-          }
-        }
-        cullingResults.transparentQueue.render(camera, pass.mask, PipelineStage.Forward);
+    const { engine, scene } = camera;
+    const { background } = scene;
+
+    const internalColorTarget = this._internalColorTarget;
+    const rhi = engine._hardwareRenderer;
+    const colorTarget = camera.renderTarget || internalColorTarget;
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:156]: you shoud write flipY logic such as`const rt2DEnabled = camera.renderTarget && cubeFace == undefined;` and ` context.applyVirtualCamera(camera._virtualCamera, false);`

File: packages/core/src/texture/enums/RenderBufferDepthFormat.ts [hunk 3]
```diff
@@ -1,22 +1,24 @@
 /**
+ * @deprecated Please use `TextureFormat` instead.
+ *
  * Render buffer depth format enumeration.
  */
 export enum RenderBufferDepthFormat {
```
Comment (zhuxudong) [inline packages/core/src/texture/enums/RenderBufferDepthFormat.ts:6]: Delete this interface directly？

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 4]
```diff
@@ -166,59 +100,100 @@ export class BasicRenderPipeline {
...
-      this._drawRenderPass(context, this._renderPassArray[i], camera, cubeFace, mipLevel, ignoreClear);
+    if (independentCanvasEnabled) {
+      const viewport = camera.pixelViewport;
+      const internalColorTarget = PipelineUtils.recreateRenderTargetIfNeeded(
+        engine,
+        this._internalColorTarget,
+        viewport.width,
+        viewport.height,
+        TextureFormat.R8G8B8A8,
+        TextureFormat.Depth24Stencil8,
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:122]: use `DepthStencil` in internal canvas is better, will use 32 depth format in webgl2

File: packages/core/src/Camera.ts [hunk 5]
```diff
@@ -59,10 +64,36 @@ export class Camera extends Component {
...
+   *
+   * @defaultValue `Downsampling.TwoX`
+   */
+  opaqueTextureDownsampling: Downsampling = Downsampling.TwoX;
+
+  /**
+   * Multi-sample anti-aliasing samples.
+   *
+   * @remarks The `opaqueTextureEnabled` property should be `true` to take effect, otherwise it will be invalid.
+   */
```
Comment (zhuxudong) [inline packages/core/src/Camera.ts:94]: this comment is error,  this parameter is used to match internal canvas

File: packages/core/src/RenderPipeline/PipelineUtils.ts [hunk 6]
```diff
@@ -62,24 +75,84 @@ export class PipelineUtils {
...
       ? PipelineUtils.recreateTextureIfNeeded(engine, currentColorTexture, width, height, colorFormat, mipmap)
       : null;
 
-    const depthTexture = depthFormat
-      ? PipelineUtils.recreateTextureIfNeeded(engine, currentDepthTexture, width, height, depthFormat, mipmap)
-      : null;
+    if (needDepthTexture) {
+      const currentDepthTexture = <Texture2D>currentRenderTarget?.depthTexture;
+      const needDepthTexture = depthFormat
+        ? PipelineUtils.recreateTextureIfNeeded(engine, currentDepthTexture, width, height, depthFormat, mipmap)
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/PipelineUtils.ts:90]: have `gc` bug, should destroy(true)

File: packages/core/src/RenderPipeline/PipelineUtils.ts [hunk 7]
```diff
@@ -62,24 +75,84 @@ export class PipelineUtils {
...
+        : null;
 
-    if (currentColorTexture !== colorTexture || currentDepthTexture !== depthTexture) {
-      currentRenderTarget?.destroy();
-      currentRenderTarget = new RenderTarget(engine, width, height, colorTexture, depthTexture);
-      currentRenderTarget.isGCIgnored = true;
+      if (currentColorTexture !== colorTexture || currentDepthTexture !== needDepthTexture) {
+        currentRenderTarget?.destroy();
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/PipelineUtils.ts:94]: `gc` bug also, see above

File: packages/core/src/RenderPipeline/PipelineUtils.ts [hunk 8]
```diff
@@ -62,24 +75,84 @@ export class PipelineUtils {
...
+        currentRenderTarget = new RenderTarget(engine, width, height, colorTexture, needDepthTexture, antiAliasing);
+        currentRenderTarget.isGCIgnored = true;
+      }
+    } else {
+      const needDepthFormat = depthFormat;
+
+      if (currentColorTexture !== colorTexture || currentRenderTarget?._depthFormat !== needDepthFormat) {
+        currentRenderTarget?.destroy();
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/PipelineUtils.ts:102]: `gc` bug also, see above

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 9]
```diff
@@ -138,87 +72,131 @@ export class BasicRenderPipeline {
...
+        false,
+        camera.msaaSamples
+      );
+      const colorTexture = internalColorTarget.getColorTexture(0);
+      colorTexture.wrapModeU = colorTexture.wrapModeV = TextureWrapMode.Clamp;
+      this._internalColorTarget = internalColorTarget;
+    } else {
+      const internalColorTarget = this._internalColorTarget;
+      if (internalColorTarget) {
+        internalColorTarget.destroy();
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:128]: will gc crash when switch `renderTarget` and open `opaqueTextureEnabled`

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts [hunk 10]
```diff
@@ -138,87 +72,131 @@ export class BasicRenderPipeline {
...
     }
 
-    pass.postRender(camera, cullingResults.opaqueQueue, cullingResults.alphaTestQueue, cullingResults.transparentQueue);
+    transparentQueue.render(camera, PipelineStage.Forward);
+
+    colorTarget?._blitRenderTarget();
+    colorTarget?.generateMipmaps();
+
+    if (internalColorTarget) {
+      PipelineUtils.blitTexture(engine, <Texture2D>internalColorTarget.getColorTexture(0), null, 0, camera.viewport);
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/BasicRenderPipeline.ts:198]: the blit target maybe `camera.renderTarget`,not always null

File: packages/core/src/RenderPipeline/DepthOnlyPass.ts [hunk 11]
```diff
@@ -56,8 +57,8 @@ export class DepthOnlyPass extends PipelinePass {
     rhi.viewport(0, 0, renderTarget.width, renderTarget.height);
     rhi.scissor(0, 0, renderTarget.width, renderTarget.height);
 
-    cullingResults.opaqueQueue.render(camera, Layer.Everything, PipelineStage.DepthOnly);
-    cullingResults.alphaTestQueue.render(camera, Layer.Everything, PipelineStage.DepthOnly);
+    cullingResults.opaqueQueue.render(camera, PipelineStage.DepthOnly);
```
Comment (zhuxudong) [inline packages/core/src/RenderPipeline/DepthOnlyPass.ts:60]: `clearRenderTarget` here after viewport and scissor

**Comment by zhuxudong** [comment]
do we should flip viewport when context.flipProjection is true ?
image

File: e2e/case/camera-opaque-texture.ts [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+      updateForE2E(engine);
+
+      initScreenshot(engine, camera);
+    });
+});
+
+function showDepthPlane(engine: Engine, camera: Entity): void {
+  const entity = camera.createChild("Plane");
+  entity.transform.setPosition(0, 0, -1);
+  entity.transform.rotate(new Vector3(-90, 0, 0));
```
Comment (zhuxudong) [inline e2e/case/camera-opaque-texture.ts:62]: rotate 90 and close `doubleFace` to test `flipY`

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts (hunk 1 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts:104]: Done

File: e2e/case/camera-opaque-texture.ts [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+      showDepthPlane(engine, cameraEntity);
+
+
+      updateForE2E(engine);
+
+      initScreenshot(engine, camera);
+    });
+});
+
+function showDepthPlane(engine: Engine, camera: Entity): void {
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply e2e/case/camera-opaque-texture.ts:59]: done

File: packages/core/src/shaderlib/extra/Blit.fs.glsl [hunk 14]
```diff
@@ -0,0 +1,15 @@
+uniform mediump sampler2D renderer_BlitTexture;
```
[the remark this answers is not in the document]
Comment (GuoLei1990) [reply packages/core/src/shaderlib/extra/Blit.fs.glsl:1]: Specifying appropriate precision can improve performance

File: packages/core/src/texture/enums/RenderBufferDepthFormat.ts (hunk 3 above)
Comment (GuoLei1990) [reply packages/core/src/texture/enums/RenderBufferDepthFormat.ts:6]: We don't want break change

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts (hunk 4 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts:117]: Depth24Stencil8  is enough，DepthStencil may improve performance cost

File: packages/core/src/Camera.ts (hunk 5 above)
Comment (GuoLei1990) [reply packages/core/src/Camera.ts:94]: This is a public API

Comment (GuoLei1990) [reply packages/core/src/Camera.ts:94]: done

File: packages/core/src/RenderPipeline/BasicRenderPipeline.ts (hunk 10 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/BasicRenderPipeline.ts:198]: If has camera.renderTarget, this func will not called

File: packages/core/src/RenderPipeline/DepthOnlyPass.ts (hunk 11 above)
Comment (GuoLei1990) [reply packages/core/src/RenderPipeline/DepthOnlyPass.ts:60]: You mean clearRenderTarget should after viewport and scissor?

File: e2e/case/camera-opaque-texture.ts (hunk 12 above)
Comment (GuoLei1990) [reply e2e/case/camera-opaque-texture.ts:62]: done

**Comment by GuoLei1990** [comment]
> do we should flip viewport when context.flipProjection is true ? image
> 
> 

Yes,has BUG!!!

**Comment by GuoLei1990** [comment]
image
I check TextureCube is inverse also!!! @zhuxudong

**Review by zhuxudong — APPROVED** [review]

**Pull request merged**